gdal/frmts/wms/gdalwmsrasterband.cpp

```diff
@@ -903,11 +909,26 @@ CPLErr GDALWMSRasterBand::ZeroBlock(int x, int y, int to_buffer_band, void *buff
                         }
                     }
                 }
-            }
+            } 
             if (p != nullptr) {
-                unsigned char *paby = reinterpret_cast<unsigned char *>(p);
-                int block_size = nBlockXSize * nBlockYSize * (GDALGetDataTypeSize(eDataType) / 8);
-                for (int i = 0; i < block_size; ++i) paby[i] = 0;
+                int hasNDV;
+                double valNDV = band->GetNoDataValue(&hasNDV);
+                if (!hasNDV)
+                    valNDV = 0;
+#define NDVSET(T) valset(reinterpret_cast<T *>(p), static_cast<T>(valNDV), block_size)
+                    switch (eDataType) {
```

Comment:
indentation is wrong, and instead of doing manual copy, use something along GDALCopyWords(&valNDV, GDT_Float64, 0, p, eDataType, GDALGetDataSizeBytes(eDataType), block_size)